File: packages/api-client-core/test/engine.test.ts
/**
 * Copyright 2026 CodeSetu Contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 */

import http from "node:http";
import type { AddressInfo } from "node:net";

import { afterAll, beforeAll, describe, expect, it } from "vitest";

interface EchoResponse {
  method: string;
  auth: string | null;
  body: string;
}

import { applyAuth } from "../src/engine/auth.js";
import { executeHttpRequest } from "../src/engine/http.js";
import { resolveVariables } from "../src/engine/variables.js";
import { createDefaultHttpRequest } from "../src/factory.js";

describe("resolveVariables", () => {
  it("applies environment precedence over collection and globals", () => {
    const out = resolveVariables("{{base}}/{{path}}", {
      globals: [{ key: "base", value: "https://global", enabled: true }],
      collection: [{ key: "base", value: "https://collection", enabled: true }],
      environment: [
        { key: "base", value: "https://env", enabled: true },
        { key: "path", value: "users", enabled: true },
      ],
    });
    expect(out).toBe("https://env/users");
  });

  it("resolves nested variables and ignores unknown tokens", () => {
    const out = resolveVariables("{{full}}", {
      environment: [
        { key: "full", value: "{{host}}:{{port}}", enabled: true },
        { key: "host", value: "localhost", enabled: true },
        { key: "port", value: "8080", enabled: true },
      ],
    });
    expect(out).toBe("localhost:8080");
    expect(resolveVariables("{{missing}}", {})).toBe("{{missing}}");
  });
});

describe("applyAuth", () => {
  it("builds a basic auth header", () => {
    const result = applyAuth({ type: "basic", basic: { username: "user", password: "pass" } });
    expect(result.headers[0]).toEqual({
      key: "Authorization",
      value: `Basic ${Buffer.from("user:pass").toString("base64")}`,
    });
  });

  it("places an api key in the query when configured", () => {
    const result = applyAuth({
      type: "apikey",
      apikey: { key: "token", value: "secret", location: "query" },
    });
    expect(result.queryParams).toEqual([{ key: "token", value: "secret" }]);
    expect(result.headers).toEqual([]);
  });
});

describe("executeHttpRequest", () => {
  let server: http.Server;
  let baseUrl = "";

  beforeAll(async () => {
    server = http.createServer((req, res) => {
      if (req.url === "/redirect") {
        res.writeHead(302, { location: "/target" });
        res.end();
        return;
      }
      if (req.url === "/target") {
        res.writeHead(200, { "content-type": "application/json" });
        res.end(JSON.stringify({ ok: true }));
        return;
      }
      const chunks: Buffer[] = [];
      req.on("data", (chunk: Buffer) => chunks.push(chunk));
      req.on("end", () => {
        res.writeHead(200, { "content-type": "application/json" });
        res.end(
          JSON.stringify({
            method: req.method,
            auth: req.headers.authorization ?? null,
            body: Buffer.concat(chunks).toString("utf-8"),
          }),
        );
      });
    });
    await new Promise<void>((resolve) => server.listen(0, resolve));
    const address = server.address() as AddressInfo;
    baseUrl = `http://127.0.0.1:${address.port}`;
  });

  afterAll(async () => {
    await new Promise<void>((resolve, reject) =>
      server.close((error) => (error ? reject(error) : resolve())),
    );
  });

  it("executes a GET and decodes a JSON body", async () => {
    const response = await executeHttpRequest(
      createDefaultHttpRequest({ method: "GET", url: `${baseUrl}/echo` }),
    );
    expect(response.status).toBe(200);
    expect(response.ok).toBe(true);
    expect((JSON.parse(response.bodyText) as EchoResponse).method).toBe("GET");
    expect(response.timings.durationMs).toBeGreaterThanOrEqual(0);
  });

  it("sends a raw JSON body with bearer auth", async () => {
    const response = await executeHttpRequest(
      createDefaultHttpRequest({
        method: "POST",
        url: `${baseUrl}/echo`,
        auth: { type: "bearer", bearer: { token: "abc123" } },
        body: { mode: "raw", rawLanguage: "json", raw: '{"hello":"world"}' },
      }),
    );
    const parsed = JSON.parse(response.bodyText) as EchoResponse;
    expect(parsed.method).toBe("POST");
    expect(parsed.auth).toBe("Bearer abc123");
    expect(parsed.body).toBe('{"hello":"world"}');
  });

  it("follows redirects and reports the final url", async () => {
    const response = await executeHttpRequest(
      createDefaultHttpRequest({ method: "GET", url: `${baseUrl}/redirect` }),
    );
    expect(response.status).toBe(200);
    expect(response.redirected).toBe(true);
    expect(response.finalUrl.endsWith("/target")).toBe(true);
  });
});

File: apps/jetbrains-apiclient/src/test/kotlin/ai/codesetu/apiclient/EngineTest.kt
package ai.codesetu.apiclient

import ai.codesetu.apiclient.engine.AuthApplier
import ai.codesetu.apiclient.engine.HttpEngine
import ai.codesetu.apiclient.engine.VariableResolver
import ai.codesetu.apiclient.model.Auth
import ai.codesetu.apiclient.model.AuthType
import ai.codesetu.apiclient.model.BasicAuth
import ai.codesetu.apiclient.model.BearerAuth
import ai.codesetu.apiclient.model.HttpRequest
import ai.codesetu.apiclient.model.RequestBody
import ai.codesetu.apiclient.model.BodyMode
import ai.codesetu.apiclient.model.RawLanguage
import ai.codesetu.apiclient.model.Variable
import ai.codesetu.apiclient.model.VariableScope
import com.sun.net.httpserver.HttpServer
import java.net.InetSocketAddress
import java.util.Base64
import kotlin.test.AfterTest
import kotlin.test.BeforeTest
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class EngineTest {
  private lateinit var server: HttpServer
  private var baseUrl = ""

  @BeforeTest
  fun start() {
    server = HttpServer.create(InetSocketAddress("127.0.0.1", 0), 0)
    server.createContext("/redirect") { exchange ->
      exchange.responseHeaders.add("Location", "/target")
      exchange.sendResponseHeaders(302, -1)
      exchange.close()
    }
    server.createContext("/target") { exchange ->
      val body = "{\"ok\":true}".toByteArray()
      exchange.responseHeaders.add("Content-Type", "application/json")
      exchange.sendResponseHeaders(200, body.size.toLong())
      exchange.responseBody.use { it.write(body) }
    }
    server.createContext("/echo") { exchange ->
      val requestBody = exchange.requestBody.readBytes()
      val auth = exchange.requestHeaders.getFirst("Authorization") ?: "null"
      val payload =
        "{\"method\":\"${exchange.requestMethod}\",\"auth\":\"$auth\",\"body\":${quote(String(requestBody))}}"
          .toByteArray()
      exchange.responseHeaders.add("Content-Type", "application/json")
      exchange.sendResponseHeaders(200, payload.size.toLong())
      exchange.responseBody.use { it.write(payload) }
    }
    server.start()
    baseUrl = "http://127.0.0.1:${server.address.port}"
  }

  @AfterTest
  fun stop() {
    server.stop(0)
  }

  @Test
  fun resolvesEnvironmentPrecedence() {
    val out = VariableResolver.resolve(
      "{{base}}/{{path}}",
      VariableScope(
        globals = listOf(Variable("base", "https://global")),
        collection = listOf(Variable("base", "https://collection")),
        environment = listOf(Variable("base", "https://env"), Variable("path", "users")),
      ),
    )
    assertEquals("https://env/users", out)
  }

  @Test
  fun buildsBasicAuthHeader() {
    val result = AuthApplier.apply(Auth(type = AuthType.BASIC, basic = BasicAuth("user", "pass")))
    val expected = "Basic " + Base64.getEncoder().encodeToString("user:pass".toByteArray())
    assertEquals("Authorization", result.headers[0].first)
    assertEquals(expected, result.headers[0].second)
  }

  @Test
  fun executesGetRequest() {
    val response = HttpEngine().execute(
      HttpRequest(method = "GET", url = "$baseUrl/echo"),
      VariableScope(),
    )
    assertEquals(200, response.status)
    assertTrue(response.ok)
    assertTrue(response.bodyText.contains("\"method\":\"GET\""))
  }

  @Test
  fun sendsBodyWithBearerAuth() {
    val response = HttpEngine().execute(
      HttpRequest(
        method = "POST",
        url = "$baseUrl/echo",
        auth = Auth(type = AuthType.BEARER, bearer = BearerAuth("abc123")),
        body = RequestBody(mode = BodyMode.RAW, rawLanguage = RawLanguage.JSON, raw = "{\"hello\":\"world\"}"),
      ),
      VariableScope(),
    )
    assertTrue(response.bodyText.contains("\"auth\":\"Bearer abc123\""))
    assertTrue(response.bodyText.contains("hello"))
  }

  @Test
  fun followsRedirects() {
    val response = HttpEngine().execute(
      HttpRequest(method = "GET", url = "$baseUrl/redirect"),
      VariableScope(),
    )
    assertEquals(200, response.status)
    assertTrue(response.redirected)
    assertTrue(response.finalUrl.endsWith("/target"))
  }

  private fun quote(value: String): String =
    "\"" + value.replace("\\", "\\\\").replace("\"", "\\\"") + "\""
}
